include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -33,6 +37,92 @@ namespace execution
 inline namespace __dpl
 {
 
+namespace __internal
+{
+
+#if _ONEDPL_PREDEFINED_POLICIES
+struct __global_instance_tag {};
+#endif
+using _QueueFactory = sycl::queue (*)();
+
+class alignas(sycl::queue) __queue_holder
+{
+    static_assert(sizeof(sycl::queue) >= sizeof(std::pair<uintptr_t, _QueueFactory>));
+    static_assert(alignof(sycl::queue) >= alignof(std::uintptr_t));
+
+    union
+    {
+        sycl::queue __q;
+        std::pair<uintptr_t, _QueueFactory> __flag_and_factory;
+    };
+
+    bool
+    __has_queue() const
+    {
+        bool res = true;
+#if _ONEDPL_PREDEFINED_POLICIES
+        // If the first size-of-pointer bytes are zeros, we consider there is no valid queue.
+        res = (__flag_and_factory.first != 0);
+        std::atomic_signal_fence(std::memory_order_acq_rel); // to prevent possible reordering due to type punning
+#endif
+        return res;
+    }
+
+  public:
+    template <typename... _Args>
+    __queue_holder(_Args... __args) : __q(std::forward<_Args>(__args)...) {}
+
+#if _ONEDPL_PREDEFINED_POLICIES
+    // The ctor for predefined policy instances does not create a queue but stores a queue factory.
+    // The first size-of-pointer bytes - the "flag" - are nullified to indicate that there is no valid queue.
+    // Then a pointer to a factory function is stored.
+    __queue_holder(__global_instance_tag, _QueueFactory __f) : __flag_and_factory(0, __f) {}
+#endif
+
+    ~__queue_holder()
+    {
+        if (__has_queue())
+            __q.~queue();
+        else
+            __flag_and_factory.second = nullptr;
+    }
+
+    // Copy and move operations have to be explicit
+    __queue_holder(const __queue_holder& __h) : __q(__h.__get_queue()) {}
+    // predefined policies should never be moved, so the move-from one must have a queue
+    __queue_holder(__queue_holder&& __h) : __q(std::move(__h.__q)) {}
+
+    __queue_holder&
+    operator=(const __queue_holder& __h)
+    {
+        if (this != &__h)
+            __q = __h.__get_queue();
+        return *this;
+    }
+
+    __queue_holder&
+    operator=(__queue_holder&& __h)
+    {
+        if (this != &__h)
+        {
+            // predefined policies should never be moved, so the move-from one must have a queue
+            assert(h.__has_queue());
```

Comment:
How about the same `assert` in move constructor?

Reply:
I have done this exercise (declare `dpcpp_default` and 'dpcpp_fpga` as `const`) in https://github.com/oneapi-src/oneDPL/pull/1671 and waiting CI.

Reply:
> I cannot add it to the member init list, or can I? Changing the list to a placement new is possible but I do not see it worthful. Meanwhile, we can consider making the predefined policies `const` as Sergey suggested, which would make these assertions even less useful.

If you have some special static function, which receive something as param and return it, I hope you may use in in initialization list and it can contain `assert` check inside.

Reply:
OK, so it would be something like this:
```
    sycl::queue&&
    __move_queue() &&
    {
        // predefined policies should never be moved, so an rvalue holder must have a queue
        assert(h.__has_queue());
        return std::move(__q);
    }

    __queue_holder(__queue_holder&& __h) : __q(std::move(__h).__move_queue()) {}

    __queue_holder&
    operator=(__queue_holder&& __h)
    {
        if (this != &__h)
            __q = std::move(__h).__move_queue();
        return *this;
    }
```
Do we want this? :)

Reply:
My slight preference is to have assertions because it's just safer. I don't know who is going to move from "alien" global objects but if we want to protect users from themselves let's do it in both assignment and constructor.

I know that we have a PR in progress, which adds `const` to global policies. If we decide to make them `const` I would prefer to remove asserts for all move semantics.

Reply:
I think we incline to add constness to global policies as in #1671, as well as fix the specification in that regard. I believe that makes the asserts useless as a typical attempt (i.e. without `const_cast` involved) to move from a global policy will result in a compilation error.

For this reason, I am going to remove the assertion. Nevertheless, I think it makes sense to document that a program which moves or destroys `dpcpp_default` etc. has undefined behavior.